ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs

```diff
@@ -15,17 +22,137 @@ module Ouroboros.Network.PeerSelection.LedgerPeers.Type
+            accStake <-
+              poolV .: "AccStake"
+              >>= (\case
+                    Just accStake -> pure accStake
+                    Nothing    ->
+                      fail "Invalid AccStake value")
+                  . readMaybe
+            reStake <-
+              poolV .: "ReStake"
+              >>= (\case
+                    Just reStake -> pure reStake
+                    Nothing    ->
+                      fail "Invalid ReStake value")
+                  . readMaybe
+            relays <- poolV .: "Relays"
+            return (accStake, (reStake, relays))) <?> Key "BigLedgerPools"
+
+          return $ LedgerPeerSnapshotV1 (slot, bigPools')
+        _ -> fail $ "Network.LedgerPeers.Type: parseJSON: failed to parse unsupported version " <> show vNum
+    case migrateLedgerPeerSnapshot parsedSnapshot of
+      Just payload -> return $ LedgerPeerSnapshot payload
+      Nothing      -> fail "Network.LedgerPeers.Type: parseJSON: failed to migrate big ledger peer snapshot"
+
+-- | cardano-slotting provides its own {To,From}CBOR instances for WithOrigin a
+-- but to pin down the encoding for CDDL we provide a wrapper with custom
+-- instances
+--
+newtype WithOriginWrapper = WithOriginWrapper (WithOrigin SlotNo)
+
+-- | Hand cranked CBOR instances to facilitate CDDL spec
+--
+instance ToCBOR WithOriginWrapper where
+  toCBOR (WithOriginWrapper Origin) = Codec.encodeListLen 1 <> Codec.encodeWord8 0
+  toCBOR (WithOriginWrapper (At (SlotNo slot))) = Codec.encodeListLen 2 <> Codec.encodeWord8 1 <> toCBOR slot
+
+instance FromCBOR WithOriginWrapper where
+  fromCBOR = do
+    listLen <- Codec.decodeListLen
+    tag <- Codec.decodeWord8
+    case (listLen, tag) of
+      (1, 0) -> pure $ WithOriginWrapper Origin
+      (1, _) -> fail "LedgerPeers.Type: Expected tag for Origin constructor"
+      (2, 1) -> WithOriginWrapper . At . SlotNo <$> fromCBOR
+      (2, _) -> fail "LedgerPeers.Type: Expected tag for At constructor"
+      _      -> fail "LedgerPeers.Type: Unrecognized list length while decoding WithOrigin SlotNo"
+
+instance ToCBOR LedgerPeerSnapshot where
+  toCBOR (LedgerPeerSnapshotV1 (slot, pools)) =
+       Codec.encodeListLen 3
+    <> Codec.encodeWord8 1
+    <> toCBOR (WithOriginWrapper slot)
+    <> toCBOR pools'
+    where
+      pools' = map (\(AccPoolStake accPoolStake, (PoolStake reStake, relays)) -> (accPoolStake, reStake, relays)) pools
+
+instance FromCBOR LedgerPeerSnapshot where
+  fromCBOR = do
+    Codec.decodeListLenOf 3
+    version <- Codec.decodeWord8
+    case version of
+      1 -> LedgerPeerSnapshotV1 <$> do
+             (WithOriginWrapper slot, pools) <- (,) <$> fromCBOR <*> fromCBOR
+             let pools' = [(AccPoolStake accStake, (PoolStake reStake, relays))
+                          | (accStake, reStake, relays) <- pools]
+             return (slot, pools')
+      _ -> fail $ "LedgerPeers.Type: no decoder could be found for version " <> show version
```

Comment:
Have this be indexed by `LEdgerSnapshotVersion`

Reply:
I am not in particular happy about using `{To,From}CBOR` type classes. We avoided using them in other parts of the codecs (e.g. mini-protocol codecs), but if we're careful, we can avoid that.